[codebase] use align_float, protect against nulls, don't leak memory etc

johnny-bit opened the pull request:
I've tried out clang's static code analyzer and tried to see what I understand from it.

This PR fixes possible bugs/crashes found via static analysis using Clang 11

dterrahe commented on src/gui/gtk.c:
```diff
@@ -3010,6 +3010,7 @@ GdkModifierType dt_key_modifier_state()
 static void notebook_size_callback(GtkNotebook *notebook, GdkRectangle *allocation, gpointer *data)
 {
   const int n = gtk_notebook_get_n_pages(notebook);
+  g_return_if_fail(n > 0);
```
This callback is only connected when n_pages >= 2, but an extra check never hurts :-)

johnny-bit replied:
The problem with callbacks is that static analyzers can't know about when callback gets called... so "better safe than sorry" ;)

dterrahe replied:
I know! So I completely support adding this check here. Just saying that it wasn't a "real" bug is all. Placating tools so that they become more helpful is a very worthwhile goal and I applaud all your efforts to strengthen the code by using as much automatic checking as is available. It saves everybody else masses of time in chasing elusive non-reproducable bugs.

TurboGit commented on src/develop/masks/path.c:
```diff
@@ -2386,7 +2386,6 @@ static int dt_path_get_mask(dt_iop_module_t *module, dt_dev_pixelpipe_iop_t *pie
   int p0[2] = { 0 }, p1[2] = { 0 };
   float pf1[2] = { 0.0f };
   int last0[2] = { -100, -100 }, last1[2] = { -100, -100 };
-  nbp = 0;
```
ok, and in line 2264 the def of nbp can be set const IINM.

TurboGit commented on src/libs/tools/darktable.c:
```diff
@@ -48,6 +48,7 @@ typedef struct dt_lib_darktable_t
   int image_width, image_height;
   // text with logo font
   cairo_surface_t *text;
+  guint8 *text_img_buffer;
```
can you comment on this change, I don't get it.
- you create a new variable
- set t
- free it

But never use it ?

johnny-bit replied:
it's for text logo svg buffer, which leaked memory... i can make it a bit better without need for keeping buffer :)

TurboGit reviewed (changes requested):
First review, some comments.

parafin commented:
I think it would have helped to add warnings compiler produced in the commit message for changes - it might have answered some of the questions "why?".

johnny-bit commented:
And maybe come back with better solutions ;)

I'll address this later this afternoon :)

pon., 11 sty 2021 o 09:12 parafin <notifications@github.com> napisał(a):

> I think it would have helped to add warnings compiler produced in the
> commit message for changes - it might have answered some of the questions
> "why?".
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/7771#issuecomment-757692336>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRRKFMJ666LF2A6VD6OD63SZKXGVANCNFSM4V32THMA>
> .
>


-- 
Pozdrawiam,
Hubert Kowalski

TurboGit commented:
BTW there is conflicts to be resolved due to recent merging.

johnny-bit commented on src/common/history.c:
```diff
@@ -1271,7 +1271,7 @@ GList *dt_history_duplicate(GList *hist)
     const dt_dev_history_item_t *old = (dt_dev_history_item_t *)(h->data);
 
-    dt_dev_history_item_t *new = (dt_dev_history_item_t *)malloc(sizeof(dt_dev_history_item_t));
+    dt_dev_history_item_t *new = (dt_dev_history_item_t *)calloc(1, sizeof(dt_dev_history_item_t));
```
Indeed this calloc was overreaching. it was mostly about params_size tbh...

johnny-bit commented on src/common/l10n.c:
```diff
@@ -189,7 +190,7 @@ static void get_language_names(GList *languages)
     goto end;
   }
 
-  char *saved_locale = strdup(setlocale(LC_ALL, NULL));
```
Well, this is problem because strdup may return NULL in rare case when it cannot allocate memory... well... It actually may crash if setlocale returns null here, but it wouldn't since setlocale with null as param will return current already allocated locale.

johnny-bit commented:
OK, this should be ready now :) addressed all comments and reviews.

as for smart stuff - one this is merged I encourage anybody with clang willing to browse for bugs do an ASAN builds etc and for static analysis try https://codechecker.readthedocs.io/en/latest/ (i used scan-build, codechecker should have far more tools and goodies)

da-phil commented:
> OK, this should be ready now :) addressed all comments and reviews.
> 
> as for smart stuff - one this is merged I encourage anybody with clang willing to browse for bugs do an ASAN builds etc and for static analysis try https://codechecker.readthedocs.io/en/latest/ (i used scan-build, codechecker should have far more tools and goodies)

Yeah, good old ASAN has been a lot of help for me too, it kind of replaces the tedious and slow valgrind most of the times ;)

Using sanitizers now inspires me to debug the potential race condition I encountered in https://github.com/darktable-org/darktable/issues/7782 using the thread sanitizer :thinking: 

Will also have a look at the clang static code analyzer, thanks for pointing it out :+1:

TurboGit reviewed (approved):
Looks ok, thanks.
